packages/oauth-helper/src/index.ts

```diff
@@ -0,0 +1,468 @@
+   */
+  public async generateInstallUrl(options: InstallURLOptions): Promise<string> {
+    // save options in case we need to share it for handleCallback failure
+    this.installOptions = options;
+
+    // Base URL
+    let URL = 'https://slack.com/oauth/v2/authorize?';
+    if (this.authVersion === 'v1') {
+      URL = 'https://slack.com/oauth/authorize?';
+    }
+
+    // scope
+    let scopes;
+    if (options.scopes instanceof Array) {
+      scopes = options.scopes.join(',');
+    } else {
+      scopes = options.scopes;
+    }
+    URL = `${URL}scope=${scopes}`;
+
+    // generate state
+    const state = await this.stateStore.generateStateParam(new Date(), options.metadata);
+    URL = `${URL}&state=${state}`;
+
+    // client id
+    URL = `${URL}&client_id=${this.clientId}`;
+
+    // redirect uri
+    if (options.redirectUri != null) {
+      URL = `${URL}&redirect_uri=${options.redirectUri}`;
+    }
+
+    // team id
+    if (options.teamId != null) {
+      URL = `${URL}&team=${options.teamId}`;
+    }
+
+    // user scope, only available for OAuth v2
+    if (options.userScopes != null && this.authVersion === 'v2') {
+      let userScopes;
+      if (options.userScopes instanceof Array) {
+        userScopes = options.userScopes.join(',');
+      } else {
+        userScopes = options.userScopes;
+      }
+      URL = `${URL}&user_scope=${userScopes}`;
+    }
+
+    return new Promise<string>((resolve) => {
+      resolve(URL);
+    });
+  }
+
+  /**
+   * This method handles the incoming request to the callback URL.
+   * It can be used as a RequestListener in almost any HTTP server
+   * framework.
+   *
+   * Verifies the state using the stateStore, exchanges the grant in the
+   * query params for an access token, and stores token and associated data
+   * in the installationStore.
+   */
+  public async handleCallback(
+    req: IncomingMessage,
+    res: ServerResponse,
+    options?: CallbackOptions,
+  ): Promise<void> {
+    console.log('handleCallback');
+    console.log(options);
+    let parsedUrl;
+    let code: string = '';
+    let state: string = '';
+
+    if (req.url != null) {
+      parsedUrl = url.parse(req.url, true);
+      code = parsedUrl.query.code as string;
+      state = parsedUrl.query.state as string;
+    }
+    console.log('code', code);
+    console.log('state', state);
+
+    try {
+      const metadata = await this.stateStore.verifyStateParam(new Date(), state);
+      console.log(metadata);
+
+      const client = new WebClient('');
+
+      let resp;
+      let installation: Installation;
+      console.log('authVersion', this.authVersion);
+      if (this.authVersion === 'v1') {
+        // convert response type from WebApiCallResult to OAuthResponse
+        resp = await client.oauth.access({
+          code,
+          client_id: this.clientId,
+          client_secret: this.clientSecret,
+        }) as unknown as OAuthV1Response;
+
+        // TODO: need to get appID
```

Comment:
`oauth.access` doesn't return `appID`. This means that I can't do the redirect to slack on a success. Is there a api call I could make to get the `appID`?

Reply:
Thanks! This looks possible but requires us adding the `users:read` scope. I'm not sure it is worth it for us to force users to add this scope just so we can get the `appId`. Would love to hear thoughts from others. Without the appID, i can't redirect back to slack on success. Instead we would have to show a generic success page.

Reply:
>Instead we would have to show a generic success page.

As the default behavior, this is totally fine. Most apps do this. If an app would like to navigate the installer to the workspace directly, the app can do that in its implementation. This library should be open to the exntension.

Reply:
Okay. Decided not to add support for appID in v1. So no `bots.info` call